returnn/compile.py

```diff
@@ -228,12 +232,29 @@ def __init__(
         :param returnn_config: RETURNN config object
         :param checkpoint: Path to the checkpoint for export
         :param device: target device for graph creation
+        :param input_names: sequence of model input names.
+            If not specified, will automatically determine from `extern_data` when available in `returnn_config.config`.
+        :param output_names: sequence of model output names.
+            If not specified, will automatically determine from `model_outputs` when available in `returnn_config.config`.
         :param returnn_python_exe: file path to the executable for running returnn (python binary or .sh)
         :param returnn_root: file path to the RETURNN repository root folder
         """
 
         self.returnn_config = returnn_config
         self.checkpoint = checkpoint
+        input_names = (
+            list(returnn_config.config["extern_data"].keys())
+            if ("extern_data" in returnn_config.config and input_names is None)
+            else input_names
+        )
+        output_names = (
+            list(returnn_config.config["model_outputs"].keys())
+            if ("model_outputs" in returnn_config.config and output_names is None)
+            else output_names
+        )
+        self.input_names = input_names
+        self.output_names = output_names
```

Comment:
Is it valid to have only one out of `input_names`, `output_names` set? (e.g. user misconfiguration, or the config contains `extern_data` but not `model_outputs`).

Should we assert that both have the same length?

Reply:
> Is it valid to have only one out of `input_names`, `output_names` set?

You mean for the job? Or for the script? But both would be valid. If one is not set, it would be automatically determined.

> (e.g. user misconfiguration, or the config contains `extern_data` but not `model_outputs`).

This would anyway then be an error in the script, because it anyway needs them to define the shape/dims.

> Should we assert that both have the same length?

You mean the input_names len = output_names len? But this doesn't need to be the case?

Or you mean that input_names matches extern_data keys, and also output_names matches model_outputs? That is already checked by the script.

Reply:
> You mean for the job? Or for the script?

I meant for the script.

> If one is not set, it would be automatically determined.

ok, this answers my question, thanks.